inc/integrations/class-coral.php

```diff
@@ -193,4 +295,157 @@ public function build_jwt( array $credentials ): string {
 	}
+
+
+	/**
+	 * Retrieve a username record. These functions use the post_title column
+	 * for the email address and the post_excerpt column for the username.
+	 *
+	 * @param string $email    The email address of the user.
+	 * @return string The username, or a blank string if none is set.
+	 */
+	private function get_username( $email ) : string {
+		global $wpdb;
+
+		$username = $wpdb->get_var( $wpdb->prepare( 
+			"SELECT post_excerpt 
+				FROM {$wpdb->posts} 
+			WHERE 
+				post_title=%s AND
+				post_type=%s AND
+				post_status='publish'
+			LIMIT 1 ",
+			[
+				$email,
+				$this->post_type,
+			]
+		) );
+
+		return $username ?? '';
+	}
+
+	/**
+	 * Retrieve a username record's post ID. These functions use the post_title column
+	 * for the email address and the post_excerpt column for the username.
+	 *
+	 * @param string $email    The email address of the user.
+	 * @return int The post ID, or 0 if none is found.
+	 */
+	private function get_username_post_id( $email ) : int {
+		global $wpdb;
+
+		$post_id = $wpdb->get_var( $wpdb->prepare( 
+			"SELECT ID 
+				FROM {$wpdb->posts} 
+			WHERE 
+				post_title=%s AND
+				post_type=%s AND
+				post_status='publish'
+			LIMIT 1 ",
+			[
+				$email,
+				$this->post_type,
+			]
+		) );
+
+		return $post_id ?? 0;
+	}
+
+	/**
+	 * Check if a username is available. These functions use the post_title column
+	 * for the email address and the post_excerpt column for the username.
+	 *
+	 * @param string $username The username to check.
+	 * @return bool Whether the name is already in use (true) or not (false).
+	 */
+	private function username_exists( $username ) : int {
+		global $wpdb;
+
+		if ( ! $username || '' === $username ) {
+			return false;
+		}
+
+		$post_id = $wpdb->get_var( $wpdb->prepare( 
+			"SELECT ID 
+				FROM {$wpdb->posts} 
+			WHERE 
+				post_excerpt=%s AND
+				post_type=%s AND
+				post_status='publish'
+			LIMIT 1 ",
+			[
+				$username,
+				$this->post_type,
+			]
+		) ) ?? 0;
+
+		return ( 0 < $post_id );
+	}
+
+	/**
+	 * Create or update a username record. These functions use the post_title column
+	 * for the email address and the post_excerpt column for the username.
+	 *
+	 * @param string $email    The email address of the user.
+	 * @param string $username The associated username.
+	 * @param string $hash     The security hash verifying the request.
+	 * @return bool Whether the create/update operation succeeded.
+	 */
+	private function set_username( $email, $username, $hash ) : bool {
+		$key         = 'username_set_hash_' . md5( $email );
+		$stored_hash = get_transient( $key );
+
+		// Stop if the hash doesn't exist, wasn't passed, or doesn't match the one on file. 
+		if ( ! $stored_hash || ! $hash || $hash !== $stored_hash ) {
+			return false;
+		}
+
+		// Stop if the username is already taken by someone else.
+		if ( $this->username_exists( $username ) ) {
+			return false;
+		}
+
+		$args = [
```

Comment:
Same concern here as above about using the email address as the primary identifier linking our user record to a Pico user. Can we use the Pico ID instead?

Reply:
👍  To @joemcgill's suggestion. The ID's _should_ be globally unique in Pico and if a user changes their email address that value should persist.

Reply:
The email definitely works as the md5 reference, though.